### ISSUE TITLE
MQTT PUBLISH failed with status MQTTNOMemory

### ISSUE THREAD

**Issue by rtheil-growlink** [Azure/azure-iot-middleware-freertos#236]
Good morning. I seem to have run across a strange issue I've not encountered before. After some time of this controller being up and running, I'm not able to send telemetry. Can disconnect and reconnect though. As you can see in the log here, I've got a memory log entry right before a call to _SendTelemetry(). The OS task high water mark is 4088 (it's always at this level), total heap free is 67172, and minHeapFree (esp_get_minimum_free_heap_size()) is 42780. Plenty of heap to send this very small message of 177 bytes. These numbers were identical when the previous message was sent successfully. I don't understand why I'd get this error if there's plenty of free memory.

```
W (16214887) memLog: azureTaskLoop: TaskMax: 4088, HeapFree: 67172, minHeapFree: 42780
E (16214887) MQTT: MQTT PUBLISH failed with status MQTTNoMemory.
E (16214897) AZ IOT: Failed to publish telemetry: MQTT error=0x00000002
E (16214897) AzureService: checkAzureError at sendMessage->AzureIoTHubClient_SendTelemetry, result: 8
E (16214907) AzureService: Send message FAIL: {"id":"00000000000024xxxxxxxxxx","messageType":"postPipeDataPoints","message":"00000000000024xxxxxxxxxx|1655089666|0|30.2|29.6|33.7|0|3.29|3.67|3.16|79.16|80.78|79.34|79.52\n"}
```

Some research led me to the documentation [here](https://www.freertos.org/Documentation/api-ref/coreMQTT/docs/doxygen/output/html/mqtt_publish_function.html), which says that it returns "MQTTNoMemory if pBuffer is too small to hold the MQTT packet. I'm unsure why pBuffer would be too small for this, nor do I fully understand where the size of pBuffer is defined for coreMQTT. I do have a buffer of 4096 defined that is passed as the pucBuffer in AzureIoTHubClient_Init() on startup.
![image](https://user-images.githubusercontent.com/96268167/173374759-a8deeb09-ff29-4c65-a5b3-26c4be1f45b8.png)
![image](https://user-images.githubusercontent.com/96268167/173374833-7687f58b-a06b-4a11-b745-8fd92f27ed8f.png)

Also found [this thread](https://docs.microsoft.com/en-us/answers/questions/694812/mqttnomemory-after-a-period-of-time-publishing.html) where the poster states that he "had moved AzureIoTHubClient_ProcessLoop() call to a OS task to be called periodically", which is exactly what I've done. I call _SendTelemetry from our main OS task loop, then azure has its own task which is constantly calling _ProcessLoop(). Like the OP there, I do not call _ProcessLoop() from our main task, opting for the azureTask to do all the work next time around. I don't understand why this would be a problem.

**Comment by danewalton** [comment]
Are you able to dig in to the code a bit and see what exactly is returning that failure from the MQTT layer?

Off the top of my head though, I could imagine a scenario where you are sending a bunch of QOS 1 messages and the `_ProcessLoop()` isn't being called often enough. [What happens is those messages are stored in the client until they receive PUBACKs since they have QOS 1 guarantees](https://github.com/FreeRTOS/coreMQTT/blob/b313f0bdbb4530d5264ec81f776bc25b0ee2a4e5/source/core_mqtt.c#L1939-L1941). By default, the MQTT library keeps an array of `MQTT_STATE_ARRAY_MAX_COUNT` un-ack'd messages, where the return [here](https://github.com/FreeRTOS/coreMQTT/blob/b313f0bdbb4530d5264ec81f776bc25b0ee2a4e5/source/core_mqtt_state.c#L504) could be `MQTTNoMemory` if that array becomes full. In our sample, [that number is generally 10](https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/154ba9fbd280b855c6e49e6e9c9ee34c35705c0f/demos/projects/ESPRESSIF/esp32/config/core_mqtt_config.h#L49).

Since you say you have a lot of heap space, that's probably where I would start. You could try increasing the number in the array via the config or increasing the frequency of calls to `_ProcessLoop()` or maybe a combination of the two.

**Comment by rtheil-growlink** [comment]
I used the middleware sample for ESP32 to implement all this, so whatever is throwing that error would be part of that. I searched around the config and component files making that connection but didn't see anything specific there, so I assume it's what you're pointing to. Is there an easy way to check on the number of messages in that un-ack'd message array? They are all definitely Qos1 messages, but I guess I'm not seeing when a message isn't successfully delivered by _ProcessLoop() as there are no errors from the azure client until this mqtt error shows up. Maybe this is an example of where I need a callback for _SendTelemetry() so I can have a count of queued vs delivered messages. This would allow me to detect a scenario where the azure client is clearly connected, but message after message is not being delivered. _ProcessLoop() is being called as fast as possible with only a vTaskDelay(1) in between.

Here's a thought though... In this test I still had a 2 minute reconnection time period (will be 50 minutes) since I was testing the changes we discussed in the other issue where I'm having to disconnect and reconnect to renew the SAS token. I have a function called doWork (thanks azure-sdk-c, lol). This runs _ProcessLoop(), then after that if it's over 2 minutes since the last successful connection, I call this reconnect process, but only after running _ProcessLoop(). I assume that if there's un-ack'd messages in the mqtt array, and a reconnection event occurs, those messages end up in some sort of purgatory? It did take about 4.5 hours to reach this point, with 1-3 messages sent every 30 seconds. I may have created this situation in my testing, but it seems like this could still happen with a 50 minute reconnection time period so I want to try to address it so the controller doesn't get stuck this way.

**Comment by danewalton** [comment]
I guess it depends on the scenario. If you want to reconnect with a clean session, [those will be cleared out here](https://github.com/FreeRTOS/coreMQTT/blob/6b1336f662a1baa05ac04f248e6eb19ffdf88809/source/core_mqtt.c#L1602-L1608) on the connect call in coreMQTT. If you want to connect without the clean session, those will stick around as dictated by the spec.

As for notification of which are ack'd, [you can set the telemetry puback callback](https://github.com/Azure/azure-iot-middleware-freertos/blob/ccfda9716f39a70c6772b6fa22f8bf7fecacee90/source/include/azure_iot_hub_client.h#L226-L227) which will notify you of any received pubacks. That should help you track how many are waiting to be puback'd. But know that any message with QOS1 will be put in the queue. So twin or command responses would also reside there.

**Comment by danewalton** [comment]
I think this is resolved so will close for now. Please let us know if you would like it reopened.

**Comment by JedidiahPaterson** [comment]
> I guess it depends on the scenario. If you want to reconnect with a clean session, [those will be cleared out here](https://github.com/FreeRTOS/coreMQTT/blob/6b1336f662a1baa05ac04f248e6eb19ffdf88809/source/core_mqtt.c#L1602-L1608) on the connect call in coreMQTT. If you want to connect without the clean session, those will stick around as dictated by the spec.
@danewalton 
> 
> As for notification of which are ack'd, [you can set the telemetry puback callback](https://github.com/Azure/azure-iot-middleware-freertos/blob/ccfda9716f39a70c6772b6fa22f8bf7fecacee90/source/include/azure_iot_hub_client.h#L226-L227) which will notify you of any received pubacks. That should help you track how many are waiting to be puback'd. But know that any message with QOS1 will be put in the queue. So twin or command responses would also reside there.

I am looking to find out how or who manages the retransmits for messages that are not Ack'ed.
What is the timeout for non-acked messages?
Can you point me to where this is managed at?
Thanks
